### PR TITLE
Add Rust implementation for FlatFIT

### PR DESCRIPTION
**Pull request by segeljakt**
This is a Rust implementation of the dynamic FlatFIT algorithm based on your C++ implementation. I went with `std::vec::Vec` for storing the internal buffer, but I think it might be possible to also use `std::vec::VecDeque`.

### CODE REVIEW

**Comment by segeljakt** [comment]
It contains unsafe code in 2 places. 1) for not having to initialise elements when rescaling the vector, 2) for being able to update the buffer during queries.

**Comment by segeljakt** [comment]
I updated it now so 2) is no longer unsafe. Now it will use `RefCell` which does a small runtime check for ownership once per push/pop/query.

File: rust/src/flatfit/mod.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            if self.size < capacity / 2 {
+                self.rescale(capacity / 2)
+            }
+        }
+    }
+    fn query(&self) -> Value {
+        let mut agg = Value::identity();
+        let mut buffer = self.buffer.borrow_mut();
+        if self.size > 0 {
+            let mut tracing_indices = Vec::new();
```
Comment (scotts) [inline rust/src/flatfit/mod.rs:74]: We actually got a significant performance improvement by making the `tracing_indices` a member of the object itself. Its use didn't change, but it was able to avoid constantly constructing the temporary vector and allocating new memory for it. When it's a permanent member, it's also reusing the same memory on each query. The size of `tracing_indices` has a low bound, so it's not a large increase in memory use.

Comment (segeljakt) [reply rust/src/flatfit/mod.rs]: Very good point. It is now fixed it in the new commit. I think I should try and do this for more of the data structures.

Comment (segeljakt) [reply rust/src/flatfit/mod.rs]: In FlatFAT, the Rust implementation uses a `HashSet` to keep track of which unique parents have been updated. Time-complexity wise a set is probably as good as it gets but I'm wondering if there is a there is a data structure with less overhead for keys which are always indices. Do you have any advice? I will in any case change it to be a member.

File: rust/src/flatfit/mod.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                self.rescale(capacity / 2)
+            }
+        }
+    }
+    fn query(&self) -> Value {
+        let mut agg = Value::identity();
+        let mut buffer = self.buffer.borrow_mut();
+        if self.size > 0 {
+            let mut tracing_indices = self.tracing_indices.borrow_mut();
+            tracing_indices.clear();
```
Comment (scotts) [inline rust/src/flatfit/mod.rs:77]: Maybe an `assert!(tracing_indices.is_empty())` is better. All of the pops from the previous query _should_ have completely cleared out the vector. If `tracing_indices` is not empty at this point, I believe that would violate invariants.

File: rust/src/flatfit/mod.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        let capacity = self.buffer.borrow().capacity();
+        if self.size + 1 > capacity {
+            self.rescale(capacity * 2)
+        }
+        let mut buffer = self.buffer.borrow_mut();
+        self.size += 1;
+        let prev = self.back;
+        self.back = (self.back + 1) % self.size;
+        buffer[self.back].val = val;
+        buffer[prev].next = self.back;
```
Comment (scotts) [inline rust/src/flatfit/mod.rs:60]: Is this correct when we push into an empty window? In that case, we start out with `front == back == size == 0`. When we end, `back == 1`, `size == 1`, `buffer[1]` contains `val` but `front == 0`. If we then do a query, doesn't that mean when we'll look at the contents of `buffer[0].val`, which is still uninitialized?

In our implementation, when the window has only one element, then `front == back`. That's what starting with `back == -1` accomplishes, although it also requires the awkwardness of indexing using a signed integer (which is not allowed in Rust, for good reason).

Our implementation has a guard around assigning the next pointer at the end of inserting/pushing for the case of adding an element to an empty window. Maybe your code needs a guard around incrementing `back`? That is, if `back == front && size == 0`, then don't increment `back`.

Comment (segeljakt) [reply rust/src/flatfit/mod.rs:60]: Hmm, I might be wrong, I thought the checks for when `size == 0` would fix it. For example:

```
Initial:
  size = 0;
  front = 0;
  back = 0;

Push:
  front = 0;
  back = if size == 0 { 0 } else { size - 1 } = 0;
  size = size + 1 = 1;
  prev = back = 0;
  back = (back + 1) % size = 0;

Query:
  agg = buffer[back] = buffer[0]

Pop:
  front = (front + 1) % 1 = 0;
  size = size - 1 = 0;
```

Comment (scotts) [reply rust/src/flatfit/mod.rs:60]: My concern is that when `size == 1`, `front == 0`, `back == 1`, is that in `query()`, I think `tracing_indices` will be `[0]` after the building loop. And then in the popping loop, we'll look at `buffer[0].val`, which is incorrect. The final statement will look at `buffer[1].val`, which is correct, but at that point, `agg` has already been built up from an unitialized value. Or am I mistaken?

It might be worthwhile to add this as a simple test - push into an empty window, query the result, check that it is correct.

File: rust/src/flatfit/mod.rs (hunk 1 above)
Comment (scotts) [reply rust/src/flatfit/mod.rs]: I think the `HashSet` is probably best. I believe you want a set to avoid putting in duplicate parents, which means the options are `HashSet` and `BTreeSet`. I believe the complexity of `BTreeSet` is worst and average-case O(log *n*) while `HashSet` is worst-case O(*n*), average-case O(1). But in this case, *n* is the number of parents of batch elements, which is guaranteed to be O(log *n*) of the window. So a worst-case O(*n*) with average-case O(1) is probably better than O(log *n*) worst and average-case.

What's important is that the size of the set is bounded by O(log *n*) of the window, which means Reactive's O(log *n*)  of the window guarantee is preserved.

As far as being specialized for `usize`, I'm not aware of anything.

File: rust/src/flatfit/mod.rs (hunk 3 above)
Comment (segeljakt) [reply rust/src/flatfit/mod.rs:60]: Hmm, I am wondering if it is ever possible for `back == 1` when `size == 1`. In either case, I added some additional tests.

Comment (scotts) [reply rust/src/flatfit/mod.rs:60]: Yes, you're correct. I was mistaken in my reasoning about `back`; I kept thinking "a small value of `back` plus one mod the size is just `back + 1`", but of course, that's not true when `back == 0` and `size == 1`.
